### PR TITLE
Adds hosts api functions

### LINKED ISSUE #62: REST and python api methods for Hosts
We dont have anything done for hosts. Need to complete that for MVP.

### PR DESCRIPTION
**Pull request by sulochan**
Resolves #62

### CODE REVIEW

**Comment by sulochan** [comment]
Note: Switched models to use custom IPAddress type rather than sqlalchemy_utils IPAddressType, this so that we can get jsonutil.to_primitive to be nice with us.

File: craton/inventory/api/v1/base.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 class Resource(restful.Resource):
     method_decorators = [request_validate, response_filter]
+#    method_decorators = [request_validate]
```
Comment (ChristopherMSpencer) [inline craton/inventory/api/v1/base.py]: what is the decision between comment vs todo?

File: craton/inventory/api/v1/hosts.py [hunk 2]
```diff
@@ -1,43 +1,154 @@
...
 
 class Hosts(base.Resource):
 
     def get(self):
-        print(g.args)
-
-        return [], 200, None
+        """
+        Get all hosts for region/cell. You can get hosts for
+        a particulr region or for a region/cell combination.
```
Comment (ChristopherMSpencer) [inline craton/inventory/api/v1/hosts.py]: sp: particular

File: craton/inventory/api/v1/base.py (hunk 1 above)
Comment (sulochan) [inline craton/inventory/api/v1/base.py]: Yeah this should not be there. Will fix.

File: craton/inventory/api/v1/hosts.py (hunk 2 above)
Comment (sulochan) [inline craton/inventory/api/v1/hosts.py]: Thanks. Will fix.

**Comment by sulochan** [comment]
Move ip_address filed to devices model

- This commit also changes the IPAddressType to
use netaddr instead of sqlalchemy ipaddress/ipaddr. This is
because netaddr is supported by jsonutils.to_premitive
and others are not

File: craton/inventory/api/v1/schemas.py [hunk 3]
```diff
@@ -2,23 +2,23 @@
...
+                   'required': ['name',
+                                'region_id',
+                                'project_id',
+                                'ip_address'],
                    'type': 'object',
                    'properties': {
                        'note': {'type': 'string'},
-                       'service': {'items': {'type': 'string'},
-                                   'type': 'array'},
                        'ip_address': {'type': 'string'},
```
Comment (sigmavirus24) [inline craton/inventory/api/v1/schemas.py]: I thought JSON Schema defined a type for ip addresses... I was using `ipv4` and `ipv6` in a different project: https://github.com/sigmavirus24/schema-validator/blob/master/examples/os-ansible-deployment/kilo/openstack_user_config_schema.yml#L46

File: craton/inventory/api/v1/schemas.py [hunk 4]
```diff
@@ -2,23 +2,23 @@
...
-                       'hostname': {'type': 'string'},
+                       'name': {'type': 'string'},
                        'id': {'type': 'integer'},
-                       'cell': {'type': 'string'},
-                       'hw_manufacturer': {'type': 'string'},
+                       'cell_id': {'type': 'string'},
+                       'project_id': {'type': 'string'},
+                       'lables': {'type': 'allOf',
```
Comment (sigmavirus24) [inline craton/inventory/api/v1/schemas.py]: lables or labels?

File: craton/inventory/api/v1/schemas.py (hunk 3 above)
Comment (sulochan) [inline craton/inventory/api/v1/schemas.py]: So I dont think there is support for anyOf in swagger. I will double check but I couldn't find it in specification last time i looked. We are using swagger to model the specification etc.

File: craton/inventory/api/v1/schemas.py (hunk 4 above)
Comment (sulochan) [inline craton/inventory/api/v1/schemas.py]: :( Will fix this.

File: craton/inventory/api/v1/hosts.py [hunk 5]
```diff
@@ -1,43 +1,139 @@
...
+            filters["host_id"] = host_id
+        if hostname != 'None':
+            filters["hostname"] = hostname
+        if ip_address != 'None':
+            filters["ip_address"] = ip_address
+        if cell != 'None':
+            filters["cell"] = cell
+
+        # This is a query constraint, you cant fetch all hosts
+        # for all regions, you have to query hosts by region.
```
Comment (mwturvey) [inline craton/inventory/api/v1/hosts.py]: Out of curiosity, why do we force this requirement?

File: craton/inventory/api/v1/schemas.py [hunk 6]
```diff
@@ -79,39 +79,34 @@
...
-                     'service': {
-                         'description': 'Openstack service to query host by',
-                         'default': [],
-                         'items': {'type': 'string'},
-                         'type': 'array'},
                      'region': {
                          'default': 'None',
                          'type': 'string',
-                         'description': 'name of the region to get hosts for'},
+                         'description': 'ID region to get hosts for'},
```
Comment (mwturvey) [inline craton/inventory/api/v1/schemas.py]: To better match the other descriptions:
ID **of the** region **for the hosts to get**

File: craton/inventory/api/v1/schemas.py [hunk 7]
```diff
@@ -79,39 +79,34 @@
...
                      'cell': {
                          'default': 'None',
                          'type': 'string',
-                         'description': 'name of the cell to get hosts for'},
+                         'description': 'ID the cell to get hosts for'},
```
Comment (mwturvey) [inline craton/inventory/api/v1/schemas.py]: ID **of the** cell **for the hosts to get**

File: craton/inventory/db/api.py [hunk 8]
```diff
@@ -98,3 +98,47 @@ def regions_data_update(context, region_id, data):
...
+# Hosts
+
+
+def hosts_get_by_region_cell(context, region_id, cell_id, filters):
+    """Get all hosts for region/cell."""
+    return IMPL.hosts_get_by_region_cell(context, region_id, cell_id, filters)
+
+
+def hosts_get_by_region(context, region_id, filters):
+    """Geta all hosts for this region."""
```
Comment (mwturvey) [inline craton/inventory/db/api.py]: Get, not Geta

File: craton/inventory/db/api.py [hunk 9]
```diff
@@ -98,3 +98,47 @@ def regions_data_update(context, region_id, data):
...
+
+
+def hosts_delete(context, host_id):
+    """Delete an existing host."""
+    return IMPL.hosts_delete(context, host_id)
+
+
+def hosts_data_update(context, host_id, data):
+    """
+    Update existing host variables or create when its not present.
```
Comment (mwturvey) [inline craton/inventory/db/api.py]: Recommend:
Update existing host variables or create them when not present.

File: craton/inventory/db/sqlalchemy/api.py [hunk 10]
```diff
@@ -288,3 +288,120 @@ def regions_data_delete(context, region_id, data):
                 pass
 
     return region_ref
+
+
+def hosts_get_by_region(context, region_id, filters):
+    """Get all hosts for this region.
+
+    :param region_id: ID for the region
+    :param filters: filters wich contails differnt keys/values to match.
```
Comment (mwturvey) [inline craton/inventory/db/sqlalchemy/api.py]: filters **which** **contain** **different** keys/values to match

**Comment by mwturvey** [comment]
A few optional suggestions and small typos, but other than those, +1 from me.

File: craton/inventory/api/v1/schemas.py (hunk 3 above)
Comment (sigmavirus24) [inline craton/inventory/api/v1/schemas.py]: Ah yeah. I always blur the line between JSON Schema that represents a thing (i.e., http://json-schema.org/latest/json-schema-core.html) and JSON Schema used to validate a thing (i.e., http://json-schema.org/latest/json-schema-validation.html). Sorry about that.

File: .travis.yml [hunk 11]
```diff
@@ -1,7 +1,14 @@
...
-python:
-  - "2.7"
-  - "3.4"
+
+matrix:
+    include:
+        - python: 2.7
+          env: TOXENV=py27
+        - python: 3.4
+          env: TOXENV=py34
```
Comment (sigmavirus24) [inline .travis.yml]: Should probably add

```yaml
        - python: 3.5
          TOXENV=py35
```

File: craton/inventory/api/v1/hosts.py [hunk 12]
```diff
@@ -1,43 +1,139 @@
...
+        region = g.args["region"]
+        cell = g.args["cell"]
+        hostname = g.args["name"]
+        host_id = g.args["id"]
+        ip_address = g.args["ip"]
+
+        context = request.environ.get("context")
+
+        filters = {}
+        if host_id != 'None':
```
Comment (sigmavirus24) [inline craton/inventory/api/v1/hosts.py]: If each of these is not set in the query string, then `g` parses it to `'None'`? That's really bizarre. Even for Flask.

File: craton/inventory/api/v1/hosts.py [hunk 13]
```diff
@@ -1,43 +1,139 @@
...
+            LOG.error("Error during host data update: %s" % err)
+            return self.error_response(500, 'Unknown Error')
 
-        return None, 400, None
+        return None, 200, None
 
     def delete(self, id):
+        """Delete host  data."""
+        # NOTE(sulo): this is not that great. Find a better way to do this.
+        # We can pass multiple keys suchs as key1=one key2=two etc. but not
```
Comment (sigmavirus24) [inline craton/inventory/api/v1/hosts.py]: Okay, don't murder me just yet, by have we considered using JSON PATCH for this use case? Then it's an update with a list of data to remove. It's not terribly bad to implement given there's already a library to do it in global-requirements. It's fairly well-supported elsewhere.

I don't presently have a concrete idea of the API in my head right now though, so maybe it wouldn't be plausible.

File: craton/inventory/db/sqlalchemy/types.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
+"""Custom SQLAlchemy types for craton inventory."""
+
+import netaddr
+import six
+from sqlalchemy import types
+
+
+class IPAddressType(types.TypeDecorator):
```
Comment (jimbaker) [inline craton/inventory/db/sqlalchemy/types.py]: maybe add TODO that this is a workaround until ipaddress support is added to oslo.serialization

File: craton/inventory/api/v1/hosts.py (hunk 12 above)
Comment (sulochan) [inline craton/inventory/api/v1/hosts.py]: Thanks for catching this! I meant to fix this. Its not flask .. its the schema validation where i had defaults as 'None'. Fixed for host. Will do the same for others in a different patch.

File: craton/inventory/api/v1/hosts.py (hunk 13 above)
Comment (sulochan) [inline craton/inventory/api/v1/hosts.py]: sound good. will take a look.

File: craton/inventory/db/sqlalchemy/types.py (hunk 14 above)
Comment (sulochan) [inline craton/inventory/db/sqlalchemy/types.py]: Done.

File: craton/inventory/api/v1/hosts.py (hunk 13 above)
Comment (sulochan) [inline craton/inventory/api/v1/hosts.py]: https://github.com/rackerlabs/craton/issues/79 for now.

File: craton/inventory/db/api.py (hunk 8 above)
Comment (sulochan) [inline craton/inventory/db/api.py]: done

File: craton/inventory/db/sqlalchemy/api.py (hunk 10 above)
Comment (sulochan) [inline craton/inventory/db/sqlalchemy/api.py]: done

File: craton/inventory/api/v1/hosts.py (hunk 13 above)
Comment (jimbaker) [inline craton/inventory/api/v1/hosts.py]: Makes sense to defer for now to #79. JSON PATCH looks like a great optimization for this case however. Also for other usages in our system.

File: .travis.yml (hunk 11 above)
Comment (jimbaker) [inline .travis.yml]: Yes, and remove 2.7, since no objections on the mailing list to removing (we can wait a few more days however). But let's do that and this change in a separate PR to keep this specific PR small/contained.

File: craton/inventory/db/sqlalchemy/api.py [hunk 15]
```diff
@@ -288,3 +288,120 @@ def regions_data_delete(context, region_id, data):
...
+
+    if "name" in filters:
+        query = query.filter_by(name=filters["name"])
+    if "ip_address" in filters:
+        query = query.filter_by(ip_address=filters["ip_address"])
+    if "id" in filters:
+        query = query.filter_by(id=filters["id"])
+    if "cell" in filters:
+        query = query.filter_by(cell_id=filters["cell"])
+
```
Comment (jimbaker) [inline craton/inventory/db/sqlalchemy/api.py]: Some code duplication here, this looks like a general query modification we want to pull out (and test)

File: craton/inventory/db/sqlalchemy/api.py [hunk 16]
```diff
@@ -288,3 +288,120 @@ def regions_data_delete(context, region_id, data):
...
+        raise exceptions.UnknownException(message=err)
+    return result
+
+
+def hosts_get_by_id(context, host_id):
+    """Get details for the host with given id."""
+    host_devices = with_polymorphic(models.Device, '*')
+    query = model_query(context, host_devices, project_only=True).\
+        filter_by(id=host_id)
+    try:
```
Comment (jimbaker) [inline craton/inventory/db/sqlalchemy/api.py]: Looks like generic exception handling that would be broadly applicable to similar queries, so ideally break out. Maybe wrap with a context manager to do so?

File: tox.ini [hunk 17]
```diff
@@ -1,6 +1,6 @@
 [tox]
 minversion = 2.0
-envlist = py34-constraints,py27-constraints,pep8-constraints
+envlist = py34-constraints,py35-constraints,py27-constraints,pep8-constraints
```
Comment (jimbaker) [inline tox.ini]: Please put into a separate PR.

**Comment by jimbaker** [comment]
+1, just need to fix the minors and it can be merged in.

File: craton/inventory/db/sqlalchemy/api.py (hunk 15 above)
Comment (sulochan) [inline craton/inventory/db/sqlalchemy/api.py]: code  duplication ? you mean the way filters are done ? we can test it but i dont know how we can pull this out ?

File: .travis.yml (hunk 11 above)
Comment (sulochan) [inline .travis.yml]: I think we can just leave this here and after a few days (need to give a few days to everyone to reply back etc) then remove it though a separate PR. I am not sure if there is anything i should do on this pr ?

File: tox.ini (hunk 17 above)
Comment (sulochan) [inline tox.ini]: this is already a separate PR. This group of PR is kinda getting messy because i am tagging in too many things :)

See: https://github.com/rackerlabs/craton/pull/71/commits/a62d45533349789e6dcf393d082bfee47a04a32f

File: craton/inventory/db/sqlalchemy/api.py (hunk 16 above)
Comment (sulochan) [inline craton/inventory/db/sqlalchemy/api.py]: Maybe... i would like to do that in a separate patch. For now i just want to get proper exception handling done ... which we dont have yet. Once this PR merges .. i have a bunch of other things to better handle exceptions etc, but not on this pr though.

File: .travis.yml (hunk 11 above)
Comment (sulochan) [inline .travis.yml]: Also all this is gone when we move to gerrit which we should do asap really.

**Comment by jimbaker** [comment]
Because a lot of stuff is held up on this, consensus on #craton is to merge and address later.
